community/bolt/src/docs/dev/messaging.asciidoc

```diff
@@ -114,8 +114,15 @@ B1 01 8C 4D  79 43 6C 69  65 6E 74 2F  31 2E 30 A3
 ==== RESET
 
 The `RESET` message is a client message used to return the current session to a "clean" state.
+It will cause the session to `IGNORE` any message it is currently processing, as well as any message before `RESET` that had not yet begun processing.
```

Comment:
So, importantly, `RESET` itself is *not* processed out-of-sequence. RESET is queued and executed in the same single-threaded manner all messages are. The difference is that it "calls ahead" and asks the state machine to ignore any messages that arrive until it sees the reset. So, lets say the queue looks like this, with the state machine currently processing the bottom message:
   
    PULL_ALL
    RUN "stmt 1"
    PULL_ALL
    RUN "stmt 2"  <-- State machine is current processing this one

When we get a RESET, two things can happen: Either the `RUN` message that was processing will still be processing when the "call-ahead" happens. If this is the case, we'll get the following outcome

    RESET         > SUCCESS
    PULL_ALL      > IGNORED
    RUN "stmt 1"  > IGNORED
    PULL_ALL      > IGNORED
    RUN "stmt 2"  > FAILURE "Transaction was terminated, this happens because blah blah"

OR, the state machine has finished processing the `RUN` it was working on, but has not yet picked up the next `PULL_ALL` message:

    RESET         > SUCCESS
    PULL_ALL      > IGNORED
    RUN "stmt 1"  > IGNORED
    PULL_ALL      > IGNORED
    RUN "stmt 2"  > SUCCESS

Overall, the "call ahead" will first tell the state machine to start ignoring messages, and then interrupt any current processing.
Importantly, this "start ignoring messages" is *not* a state change. *all* state changes happen through the regular single-threaded message processing. 
This is super-duper important that we don't change, the state machine is not thread safe, and making it thread safe would be a pain.
Instead, this "start ignoring" call-ahead is done through a separate atomic integer, which acts as the only mode of communication between the thread calling ahead and the thread processing messages.